### ISSUE TITLE
no role="img" if alt-attribute empty

### ISSUE THREAD

**Issue by izaquieltome** [quellenform/t3x-iconpack#38]
Iconpack creates SVG-Icons with role="img" but without an alt-attribute, which is wrong in WCAG accessibility definitions.

An svg-icon without an alt-attribute should not use role="img" as this is bad for accessibility. 
role="img" should be used only when an alt-text is present, so it is interpreted correctly as an image. 

I did this now very dirty by filtering out the role-attribute in the php-class - but this can be done better and more stable for shure :) 

But it definitively should be considered, as for EU it's pretty important - and especially here in Austria starting with 28th of June accessibility in Websites (and all digital services) becomes a law.   

Thanks for your help,
Iza

**Comment by stephankellermayr** [comment]
Thank you for your feedback on this issue! However, this is not a problem with the iconpack extension, but rather with the respective iconpack. The configuration file for the iconpack may contain instructions to add this attribute.

This would therefore need to be changed in the respective iconpack or in its [configuration file](https://github.com/quellenform/t3x-iconpack-fontawesome/blob/main/Configuration/Iconpack/FontAwesome6.yaml).

Which iconpack did you notice this in, or which iconpack are you using?

**Comment by stephankellermayr** [comment]
...However, I can also imagine making a change in the Extension Iconpack so that if no alt or title attribute is set, the role attribute is also removed.

Would that be a solution?

**Comment by izaquieltome** [comment]
I recognized it by using the fontawesome iconpack. But now also checked "elegent", "lineicons", "themify" and "typo3" iconpacks (al of them quellenform-iconpacks) - they all use "role = img" in their yaml-configuration.
So maybe it would be a "cleaner" solution to remove the role attribute too, when no alt or title is set and keep the attribute, when either alt or title is set?
 
So every iconpack can use the role="img" attribute for icons which are used in an image-context (with alt and/or title), and to remove the role-attribute, when neither of the two attributes is set?

**Comment by stephankellermayr** [comment]
Currently, attributes are defined by the respective iconpacks.
The iconpack extension adds an empty `alt` attribute unless the editor or developer has explicitly defined an additional `alt` attribute.
I think this is correct (please correct me if I am wrong).

But to fully understand the issue, please help me with the following assumption:

**webfont & svg**
An `alt` attribute should always be used for icons of type *webfont* and *svg*. If none has been explicitly defined by the editor or developer, an empty `alt` attribute should be added.

**svgInline & svgSprite**
If available, `alt` and `title` attributes are written directly in the HTML (`<desc>` and `<title>`).

In all cases, however, a `role="img"` attribute should only be used if either `alt` and/or `title` attributes are present.

If no `alt` or `title` attributes are present (even if `alt` is empty), then the iconpack extension should remove the `role` attribute.

Is that correct?

**Comment by stephankellermayr** [comment]
@izaquieltome One more thought on the topic of accessibility/screen readers:  
Wouldn't it make sense to add an `aria-hidden="true"` attribute (possibly optional and controlled via TypoScript), which would cause icons to be ignored by screen readers in general?

As much as I like icons, I don't think they offer any added value for people with disabilities.

Am I wrong here?

**Comment by izaquieltome** [comment]
@stephankellermayr , I think you are absolutely right, an aria-hidden="true" would also make sense.

For the role-Attribute:
An Icon should never use a role="img" attribute, as it does not really provide effective additional information for the user, so it does not need to be treated as an image. 
If it uses a role="img" tag, it gets treated as an image and therefore is missing additional Information like an alt or title attribute.

Therefore my suggestion, to also filter out the role-attribute when alt and title are empty. Because in very rare occasions an icon still could be used as an image with additional information to the text.

**Comment by stephankellermayr** [comment]
@izaquieltome OK, thanks, that sounds reasonable!

I will take another closer look at the scenario and aim to make the following changes:

1. The `role` and `aria-hidden` attributes will be completely removed from all icon packs!
2. *EXT:iconpack* will add the `aria-hidden` attribute to all icons (webfont, svg, svgInline, svgSprite). This will cause all icons to be ignored by screen readers by default.
3. One exception: If a `title` or `alt` attribute is explicitly used in the RTE, the `aria-hidden` attribute is not used!
3. In addition to the scopes "native" (for TYPO3 fields) and "rte" (body text), there will be a new scope: "fluid". Here, exactly what is written in the template is always used, so that you still have the option of explicitly influencing the appearance of icons added to any templates via the ViewHelper.

Does that make sense?

**Comment by stephankellermayr** [comment]
Well, I've made one mistake here: The native TYPO3 fields are rendered via the ViewHelper anyway...so the "fluid" scope doesn't make much sense.

**Comment by izaquieltome** [comment]
I think, this is a great solution :)

**Comment by stephankellermayr** [comment]
@izaquieltome Solved with [4b6c390] ...Thanks again for your feedback and commitment!

**Comment by stephankellermayr** [comment]
@izaquieltome Addendum: The icon pack update will follow shortly.
